### PR TITLE
Unify API of BudgetAccountants

### LINKED ISSUE #68: Unifying API of BudgetAccountant and PLDBudgetAccountant
# Context 
The privacy budget accounting is an important feature of the DP aggregation system. Currently there are 2 [BudgetAccoutant](https://github.com/OpenMined/PipelineDP/blob/24f9a7bd6c488e68d8297b031999525a5971e235/pipeline_dp/budget_accounting.py#L56) and [PLDBudgetAccountant](https://github.com/OpenMined/PipelineDP/blob/24f9a7bd6c488e68d8297b031999525a5971e235/pipeline_dp/budget_accounting.py#L143). 

The `BudgetAccountant` uses the naive composition, namely the total budget is the sum of budgets:

(𝜺_1, 𝛿_1) + (𝜺_2, 𝛿_2) = (𝜺_1+𝜺_2,𝛿_1+𝛿_2).

While PLDBudgetAccountant uses Privacy Loss Distribution (PLD) method for advance composition. PLD computations are implemented in [dp_accounting](https://github.com/google/differential-privacy/tree/main/python/dp_accounting) library. 

PLD is more effective from accuracy point of view (i.e. it's possible to add less noise to the result, while ensuring the same privacy guarantees), but have some limitations (eg. it might take long time for PLD computations). In future it's possible that more accounts will be implemented.

## How API of BudgetAccounant and PLDBudgetAccountant differs now
1. Both of them have `request_budget`, which have different arguments and different return types

2. Both of them `compute_budgets` methods, w/o argument and returns. But implementations are different.

The idea is to unify arguments and return of `request_budget` (and to make corresponding changes in the implementation)

# Goals
Introduce the common interface for these 2 classes, to ensure that they might be used in polymorphic way.

# Changes
1.Renaming `BudgetAccounant` -> `NaiveBudgetAccounant`

2.Introducing a base abstract class

```
@abstract
class BudgetAccounant:
   @abstractmethod
   def request_budget(self, noise_kind: NoiseKind, sensitivity: float = 1,weight: float = 1, count: int = 1) ->  MechanismSpec:

   @abstractmethod
   def request_budget():

```

3.Updating `MechanismSpec` by adding properties   

```
@dataclass
class MechanismSpec:
    ... # as now 
   @property
    def eps(self):

   @property
    def delta(self):

```

4.Making corresponding changes in implementation and tests

### PR DESCRIPTION
**Pull request by zachferr**
## Description
I unified the API for Budget Accountants and added an abstract base class. This will close #68. I also made changes to other files that were using BudgetAccountant previously.

I couldn't test my changes to dp_engine due to the failing import so I'm not sure how to make sure those changes were okay.

I'm not 100% sure on if my changes to the implementation were correct but all the tests that were passing before still passed.

## Affected Dependencies
N/A

## How has this been tested?
- Ran the same unit tests as before.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/budget_accounting.py [hunk 1]
```diff
@@ -8,139 +9,149 @@
...
 @dataclass
-class Budget:
-    """Manages the budget allocated for an operation.
+class MechanismSpec:
+    """Specifies the parameters for a mechanism.
 
-    The values for eps and delta are computed when the method compute_budgets
-    of the corresponding BudgetAccount is called.
+    NoiseKind defines the kind of noise distribution.
+    noise is the minimized noise standard deviation.
```
Comment (dvadym) [inline pipeline_dp/budget_accounting.py:16]: noise -> _noise_standard_deviation

Please add in comments
(eps, delta) are parameters of (eps, delta)-differential privacy

**Review by dvadym — COMMENTED** [review]
Thanks, it looks good! I left some comments

Please add try/except on the failing import [here](https://github.com/OpenMined/PipelineDP/blob/d1288fbe21d85afc8b94f14f103603969dea6955/pipeline_dp/dp_engine.py#L14). 
With TODO remove when PyDP is fixed.

Such way it would be possible to run most tests (at least affected tests)

**Comment by dvadym** [comment]
FYI the new version of python-dp is released ([1.11](https://pypi.org/project/python-dp/)). So no need to add try/except everything should work

File: pipeline_dp/budget_accounting.py (hunk 1 above)
Comment (zachferr) [reply pipeline_dp/budget_accounting.py]: Done.

File: pipeline_dp/budget_accounting.py [hunk 2]
```diff
@@ -8,139 +9,149 @@
...
-        return self._noise_standard_deviation
-
-
-@dataclass
-class MechanismSpecInternal:
-    """Stores sensitivity and weight not exposed in MechanismSpec."""
-    sensitivity: float
-    weight: float
-    mechanism_spec: MechanismSpec
+            if mechanism.mechanism_spec.noise_kind == NoiseKind.GAUSSIAN:
```
[the remark this answers is not in the document]
Comment (zachferr) [reply pipeline_dp/budget_accounting.py]: Done.

File: pipeline_dp/budget_accounting.py [hunk 3]
```diff
@@ -8,139 +9,149 @@
...
+                if total_weight_delta:
+                    numerator = self._total_delta * mechanism.weight
+                    delta = numerator / total_weight_delta
+            mechanism.mechanism_spec._eps = eps
```
[the remark this answers is not in the document]
Comment (zachferr) [reply pipeline_dp/budget_accounting.py]: Done.

**Comment by zachferr** [comment]
Thanks for the comments! 

test_aggregate_report is still broken in dp_engine_test but I don't believe that it is due to my changes.

File: pipeline_dp/budget_accounting.py [hunk 4]
```diff
@@ -8,139 +9,165 @@
...
-        self._requested_budgets = []
+        self._total_epsilon = total_epsilon
+        self._total_delta = total_delta
+        self._mechanisms = []
+
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        """Request a budget.
```
Comment (dvadym) [inline pipeline_dp/budget_accounting.py:120]: Nit: Request**s**

File: pipeline_dp/budget_accounting.py [hunk 5]
```diff
@@ -8,139 +9,165 @@
...
 
-    def request_budget(self, weight: float, *, use_eps: bool,
-                       use_delta: bool) -> Budget:
-        """Requests a budget.
+        Constructs a mechanism spec based on the parameters.
+        Adds the mechanism to the pipeline for future calculation.
```
Comment (dvadym) [inline pipeline_dp/budget_accounting.py:123]: Nit: Keeps the mechanism spec for future calculations.

File: pipeline_dp/budget_accounting.py [hunk 6]
```diff
@@ -8,139 +9,165 @@
...
-            AssertionError: The privacy budget is not calculated yet.
+                Raises:
+                    AssertionError: The privacy budget is not calculated yet.
         """
         if self._delta is None:
             raise AssertionError("Privacy budget is not calculated yet.")
         return self._delta
 
-    def set_eps_delta(self, eps, delta):
+    def set_eps_delta(self, eps: float, delta: float) -> None:
```
Comment (dvadym) [inline pipeline_dp/budget_accounting.py:56]: Nit: delta: Optional[float]

and no need in "delta can be None"

File: pipeline_dp/budget_accounting.py [hunk 7]
```diff
@@ -8,139 +9,165 @@
 
 
 @dataclass
-class Budget:
-    """Manages the budget allocated for an operation.
+class MechanismSpec:
+    """Specifies the parameters for a mechanism.
```
Comment (dvadym) [inline pipeline_dp/budget_accounting.py:13]: Nit: for a **DP** mechanism.

File: pipeline_dp/budget_accounting.py [hunk 8]
```diff
@@ -8,139 +9,165 @@
...
+    @abc.abstractmethod
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        pass
+
+    @abc.abstractmethod
+    def request_budget(self):
+        pass
```
Comment (dvadym) [inline pipeline_dp/budget_accounting.py:93]: Please add abstract method `compute_budgets()`

**Review by dvadym — COMMENTED** [review]
Thanks, it looks good! A few minor comments (mostly for comments)

File: pipeline_dp/budget_accounting.py (hunk 8 above)
Comment (zachferr) [reply pipeline_dp/budget_accounting.py:93]: Sorry about this I accidentally doubled this method.

**Review by dvadym — APPROVED** [review]
Thanks a lot for unifying API of budget accountants!

**Pull request merged**